kube/src/config/mod.rs

```diff
@@ -14,10 +14,60 @@ use crate::{Error, Result};
 pub use file_loader::KubeConfigOptions;
 use file_loader::{ConfigLoader, Der};
 
+use chrono::{DateTime, Utc};
 use reqwest::header::{self, HeaderMap};
+use tokio::sync::Mutex;
 
+use std::sync::Arc;
 use std::time::Duration;
 
+#[derive(Debug, Clone)]
+pub(crate) enum Authentication {
+    None,
+    Basic(String),
+    Token(String),
+    RefreshableToken(Arc<Mutex<(String, DateTime<Utc>)>>, ConfigLoader),
```

Comment:
I guess the `Arc<Mutex<_>>` is here because you have multiple queries happening at the same time, and you only want to update the token once, so the intent here is to block all queries if there's a need to refresh the token, and wait for that to complete?

Reply:
Yeah, so `Arc` is to make sure the `Config` is still `Clone` and the Mutex makes sure you only update once and to block other token queries from happening while refreshing. This was primarily to give us interior mutability without have the need for `Config` or `Client` to be `mut`

Reply:
Yeah, that makes sense. I suppose having the lock lifted to the client would actually convolute things quite a bit.